src/main/java/co/cask/hydrator/plugin/source/KafkaStreamingSource.java

```diff
@@ -0,0 +1,292 @@
+  }
+
+  @Override
+  public JavaDStream<StructuredRecord> getStream(StreamingContext context) throws Exception {
+    context.registerLineage(conf.referenceName);
+
+    Map<String, String> kafkaParams = new HashMap<>();
+    kafkaParams.put("metadata.broker.list", conf.getBrokers());
+
+    List<SimpleConsumer> consumers = new ArrayList<>();
+    for (Map.Entry<String, Integer> brokerEntry : conf.getBrokerMap().entrySet()) {
+      consumers.add(new SimpleConsumer(brokerEntry.getKey(), brokerEntry.getValue(),
+                                       20 * 1000, 128 * 1024, "partitionLookup"));
+    }
+
+    try {
+      Map<TopicAndPartition, Long> offsets = conf.getInitialPartitionOffsets(getPartitions(consumers));
+      // KafkaUtils doesn't understand -1 and -2 as smallest offset and latest offset.
+      // so we have to replace them with the actual smallest and latest
+      Map<TopicAndPartition, PartitionOffsetRequestInfo> offsetsToRequest = new HashMap<>();
+      for (Map.Entry<TopicAndPartition, Long> entry : offsets.entrySet()) {
+        TopicAndPartition topicAndPartition = entry.getKey();
+        Long offset = entry.getValue();
+        if (offset == OffsetRequest.EarliestTime() || offset == OffsetRequest.LatestTime()) {
+          offsetsToRequest.put(topicAndPartition, new PartitionOffsetRequestInfo(offset, 1));
+        }
+      }
+
+      kafka.javaapi.OffsetRequest offsetRequest =
+        new kafka.javaapi.OffsetRequest(offsetsToRequest, OffsetRequest.CurrentVersion(), "offsetLookup");
+      Set<TopicAndPartition> offsetsFound = new HashSet<>();
+      for (SimpleConsumer consumer : consumers) {
+        OffsetResponse response = consumer.getOffsetsBefore(offsetRequest);
+        for (TopicAndPartition topicAndPartition : offsetsToRequest.keySet()) {
+          String topic = topicAndPartition.topic();
+          int partition = topicAndPartition.partition();
+          if (response.errorCode(topic, partition) == 0) {
+            offsets.put(topicAndPartition, response.offsets(topic, partition)[0]);
+            offsetsFound.add(topicAndPartition);
+          }
+        }
+      }
+
+      Set<TopicAndPartition> missingOffsets = Sets.difference(offsetsToRequest.keySet(), offsetsFound);
+      if (!missingOffsets.isEmpty()) {
+        throw new IllegalStateException(String.format(
+          "Could not find offsets for %s. Please check all brokers were included in the broker list.", missingOffsets));
+      }
+      LOG.info("Using initial offsets {}", offsets);
+
+      return KafkaUtils.createDirectStream(
+        context.getSparkStreamingContext(), byte[].class, byte[].class, DefaultDecoder.class, DefaultDecoder.class,
+        MessageAndMetadata.class, kafkaParams, offsets,
+        new Function<MessageAndMetadata<byte[], byte[]>, MessageAndMetadata>() {
+          @Override
+          public MessageAndMetadata call(MessageAndMetadata<byte[], byte[]> in) throws Exception {
+            return in;
```

Comment:
KafkaUtils.createDirectStream() makes you pass in a function if you also want to pass in the starting offsets. I'm not sure why I used an identity here and then called .transform() on the DStream instead of just passing in a Function here though... should see if we can just pass in the BytesFunction or FormatFunction here (depending on the format config setting)